docs/pokesay.1
.TH POKESAY 1 "September 2025" "pokesay 0.17.0" "User Commands"
.SH NAME
pokesay \- print Pokémon in the CLI! An adaptation of the classic "cowsay"
.SH SYNOPSIS
.B pokesay
[\fIOPTIONS\fR] [\fIPARAMETERS\fR ...]
.SH DESCRIPTION
.B pokesay
takes piped text and displays it in a speech bubble spoken by a Pokémon sprite.

.SH OPTIONS
.TP
.BR \-b ", " --info-border
Draw a border around the info box.
.TP
.BR \-B ", " --no-bubble
Do not draw the speech bubble.
.TP
.BR \-c ", " --category=\fIVALUE\fR
Choose a Pokémon from a specific category.
.TP
.BR \-C ", " --no-category-info
Do not print Pokémon category information in the info box.
.TP
.BR \-f ", " --fastest
Run with the fastest possible configuration (\-\-nowrap & \-\-notabspaces).
.TP
.BR \-F ", " --flip
Flip the Pokémon horizontally (face right instead of left).
.TP
.BR \-h ", " --help
Display this help message.
.TP
.BR \-i ", " --id=\fIVALUE\fR
Choose a Pokémon from a specific ID (see \fBpokesay -l\fR for IDs).
.TP
.BR \-I ", " --id-info
Print the Pokémon ID in the info box.
.TP
.BR \-j ", " --japanese-name
Print the Japanese name in the info box.
.TP
.BR \-L ", " --list-categories
List all available categories.
.TP
.BR \-l ", " --list-names[=\fIVALUE\fR]
List all available names.
.TP
.BR \-n ", " --name=\fIVALUE\fR
Choose a Pokémon from a specific name.
.TP
.BR \-s ", " --no-tab-spaces
Do not replace tab characters (fastest).
.TP
.BR \-t ", " --tab-width=\fIVALUE\fR
Replace any tab characters with N spaces [4].
.TP
.BR \-u ", " --unicode-borders
Use unicode characters to draw the border around the speech box (and info box if \-\-info-border is enabled).
.TP
.BR \-v ", " --verbose
Print verbose output.
.TP
.BR \-W ", " --no-wrap
Disable text wrapping (fastest).
.TP
.BR \-w ", " --width=\fIVALUE\fR
The max speech bubble width [80].


.SH EXAMPLES
Pipe text to pokesay:

.EX
    echo "yolo" | pokesay
.EE

See a random fortune every time you open a terminal (requires fortune):

.EX
    echo 'fortune | pokesay' >> $HOME/.bashrc
.EE

List all available categories:

.EX
    pokesay -L
.EE

List all available names:

.EX
    pokesay -l
.EE

Print a message with a random pokemon:

.EX
    echo 'Hello, world!' | pokesay
.EE

Print a message with a specific pokemon:

.EX
    echo 'Hello, world!' | pokesay -n pikachu
.EE

Print a message with a specific pokemon category:

.EX
    echo 'Hello, world!' | pokesay -c big
    echo 'Hello, world!' | pokesay -c shiny
.EE

Print a message with a specific pokemon category and name:

.EX
    echo 'Hello, world!' | pokesay -c shiny -n charizard
.EE

Print a specific pokemon by its ID:

.EX
    echo 'Hello, world!' | pokesay -i 0491.1719
.EE

.SH FILES
None.

.SH AUTHOR
Tom McKeesick <tmck01@gmail.com>

.SH SEE ALSO
.BR cowsay (1)

.SH HOMEPAGE
https://github.com/tmck-code/pokesay



pokesay.go
package main

import (
	"embed"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"

	"github.com/pborman/getopt/v2"
	"github.com/tmck-code/pokesay/src/pokedex"
	"github.com/tmck-code/pokesay/src/pokesay"
	"github.com/tmck-code/pokesay/src/timer"
)

var (
	//go:embed build/assets/category_keys.txt
	GOBCategoryKeys []byte
	//go:embed build/assets/names.txt
	GOBAllNames []byte

	//go:embed build/assets/total.txt
	GOBTotal []byte
	//go:embed build/assets/cows/*cow
	GOBCowData embed.FS
	//go:embed build/assets/metadata/*metadata
	GOBCowNames embed.FS
	//go:embed all:build/assets/categories
	GOBCategories embed.FS

	CategoryRoot string = "build/assets/categories" // the root directory of the pokemon categories
	MetadataRoot string = "build/assets/metadata"   // the root directory of the pokemon metadata
	CowDataRoot  string = "build/assets/cows"       // the root directory of the pokemon cow data
)

// parseFlags parses the command line flags and returns a pokesay.Args struct
func parseFlags() pokesay.Args {
	help := getopt.BoolLong("help", 'h', "display this help message")
	// print verbose output (currently timer output)
	verbose := getopt.BoolLong("verbose", 'v', "print verbose output", "verbose")

	// selection/filtering
	name := getopt.StringLong("name", 'n', "", "choose a pokemon from a specific name")
	id := getopt.StringLong("id", 'i', "", "choose a pokemon from a specific ID (see `pokesay -l` for IDs)")
	category := getopt.StringLong("category", 'c', "", "choose a pokemon from a specific category")

	// list operations
	listNames := getopt.StringLong("list-names", 'l', "", "list all available names")
	getopt.Lookup('l').SetOptional()
	listCategories := getopt.BoolLong("list-categories", 'L', "list all available categories")

	width := getopt.IntLong("width", 'w', 80, "the max speech bubble width")

	// speech bubble options
	tabWidth := getopt.IntLong("tab-width", 't', 4, "replace any tab characters with N spaces")
	noWrap := getopt.BoolLong("no-wrap", 'W', "disable text wrapping (fastest)")
	noTabSpaces := getopt.BoolLong("no-tab-spaces", 's', "do not replace tab characters (fastest)")
	fastest := getopt.BoolLong("fastest", 'f', "run with the fastest possible configuration (--nowrap & --notabspaces)")
	noBubble := getopt.BoolLong("no-bubble", 'B', "do not draw the speech bubble")

	// info box options
	japaneseName := getopt.BoolLong("japanese-name", 'j', "print the japanese name in the info box")
	showId := getopt.BoolLong("id-info", 'I', "print the pokemon ID in the info box")
	noCategoryInfo := getopt.BoolLong("no-category-info", 'C', "do not print pokemon category information in the info box")
	drawInfoBorder := getopt.BoolLong("info-border", 'b', "draw a border around the info box")

	// other option
	unicodeBorders := getopt.BoolLong("unicode-borders", 'u', "use unicode characters to draw the border around the speech box (and info box if --info-border is enabled)")
	flipPokemon := getopt.BoolLong("flip", 'F', "flip the pokemon horizontally (face right instead of left)")

	getopt.Parse()
	var args pokesay.Args

	if *fastest {
		args = pokesay.Args{
			Width:       *width,
			NoWrap:      true,
			TabSpaces:   "    ",
			NoTabSpaces: true,
			BoxChars:    pokesay.DetermineBoxChars(false),
			Help:        *help,
			Verbose:     *verbose,
		}
	} else {
		args = pokesay.Args{
			Width:          *width,
			NoWrap:         *noWrap,
			DrawBubble:     !*noBubble,
			TabSpaces:      strings.Repeat(" ", *tabWidth),
			NoTabSpaces:    *noTabSpaces,
			NoCategoryInfo: *noCategoryInfo,
			ListCategories: *listCategories,
			ListNames:      getopt.GetCount("list-names") > 0,
			ListNameToken:  *listNames,
			Category:       *category,
			NameToken:      *name,
			IDToken:        *id,
			JapaneseName:   *japaneseName,
			ShowID:         *showId,
			BoxChars:       pokesay.DetermineBoxChars(*unicodeBorders),
			DrawInfoBorder: *drawInfoBorder,
			FlipPokemon:    *flipPokemon,
			Help:           *help,
			Verbose:        *verbose,
		}
	}
	return args
}

// runListCategories prints all available categories
// - This reads a list of categories from the embedded filesystem
// - prints the list of categories, and the total number of categories
func runListCategories() {
	categories := pokedex.ReadStructFromBytes[[]string](GOBCategoryKeys)
	fmt.Printf("%s\n%d %s\n", strings.Join(categories, " "), len(categories), "total categories")
}

// runListNames prints all available pokemon names
// - This reads a struct of {name -> metadata indexes} from the embedded filesystem
// - prints all the keys of the struct, and the total number of names
func runListNames(token string) {
	t := timer.NewTimer("runListNames", true)

	pokedex.ReadStructFromBytes[map[string][]int](GOBAllNames)
	namesSorted := pokedex.GatherMapKeys(pokedex.ReadStructFromBytes[map[string][]int](GOBAllNames))
	t.Mark("read metadata")

	s := make(map[string]map[string]string)
	exit := false
	for i, name := range namesSorted {
		metadata := pokedex.ReadMetadataFromEmbedded(GOBCowNames, pokedex.MetadataFpath(MetadataRoot, i))

		entries := make(map[string]string, 0)

		for _, entry := range metadata.Entries {
			k := fmt.Sprintf("%04d.%04d", i, entry.EntryIndex)
			entries[k] = strings.Join(entry.Categories, ", ")
		}
		if token != "" && token == name {
			json, _ := json.MarshalIndent(map[string]map[string]string{name: entries}, "", strings.Repeat(" ", 2))
			fmt.Fprintln(os.Stdout, string(json))
			exit = true
			break
		}
		s[name] = entries
	}
	t.Mark("read metadata")
	if exit {
		return
	}
	json, _ := json.MarshalIndent(s, "", strings.Repeat(" ", 2))
	fmt.Fprintln(os.Stdout, string(json))
}

// GenerateNames returns a list of names to print
// - If the japanese name flag is set, it returns both the english and japanese names
// - Otherwise, it returns just the english name
func GenerateNames(metadata pokedex.PokemonMetadata, args pokesay.Args, final pokedex.PokemonEntryMapping) []string {
	nameParts := []string{metadata.Name}
	if args.JapaneseName {
		nameParts = append(nameParts, fmt.Sprintf("%s (%s)", metadata.JapaneseName, metadata.JapanesePhonetic))
	}
	if args.ShowID {
		nameParts = append(nameParts, fmt.Sprintf("%s.%04d", metadata.Idx, final.EntryIndex))
	}
	return nameParts
}

// runPrintByName prints a pokemon matched by a name
// The name must match the lowercase name of the pokemon (TODO: improve this behaviour)
// - This reads a struct of {name -> metadata indexes} from the embedded filesystem
// - It matches the name to a metadata index, loads the corresponding metadata file, and then chooses a random entry
// - Finally, it prints the pokemon
func runPrintByName(args pokesay.Args) {
	names := pokedex.ReadStructFromBytes[map[string][]int](GOBAllNames)
	timer.DebugTimer.Mark("read name struct")

	metadata, final := pokesay.ChooseByName(names, args.NameToken, GOBCowNames, MetadataRoot)

	pNames := GenerateNames(metadata, args, final)
	timer.DebugTimer.Mark("generate names")

	pokesay.Print(args, final.EntryIndex, pNames, final.Categories, GOBCowData)
}

// runPrintByID prints a pokemon corresponding to a specific ID
// - This reads a list of names from the embedded filesystem
// - It finds the name at alphabetical index `IDToken`
// - It matches the name to a metadata index, loads the corresponding metadata file, and then chooses a random entry
// - Finally, it prints the pokemon
func runPrintByID(args pokesay.Args) {
	idxs := strings.Split(args.IDToken, ".")

	idx, _ := strconv.Atoi(idxs[0])
	subIdx, _ := strconv.Atoi(idxs[1])

	timer.DebugTimer.Mark("format IDs")

	metadata, final, err := pokesay.ChooseByIndex(idx, subIdx, GOBCowNames, MetadataRoot)
	if err != nil {
		log.Fatal(err)
	}

	pokesay.Print(args, subIdx, GenerateNames(metadata, args, final), final.Categories, GOBCowData)
}

// runPrintByCategory prints a pokemon matched by a category
// - This loads a GOB file containing a pokemon "category" search struct from the embedded filesystem
// - It chooses a random category file from the corresponding category directory
// - It reads the category file and chooses a random pokemon from the category
//   - # TODO: as each category dir contains files with a singular entry of {metadata index/entry index},
//   - # this means that pokemon that are in the same category multiple times will be chosen more often
//
// - It reads the metadata file of the chosen pokemon and chooses the corresponding entry from the category search
// - Finally, it prints the pokemon
func runPrintByCategory(args pokesay.Args) {
	dirPath := pokedex.CategoryDirpath(CategoryRoot, args.Category)
	dir, _ := GOBCategories.ReadDir(dirPath)
	metadata, final := pokesay.ChooseByCategory(args.Category, dir, GOBCategories, CategoryRoot, GOBCowNames, MetadataRoot)

	pokesay.Print(args, final.EntryIndex, GenerateNames(metadata, args, final), final.Categories, GOBCowData)
}

// runPrintByNameAndCategory prints a pokemon matched by a name and category
// - This reads a struct of {name -> metadata indexes} from the embedded filesystem
// - It matches the name to a metadata index, loads the corresponding metadata file, and then randomly chooses an entry that matches the category
// - Finally, it prints the pokemon
func runPrintByNameAndCategory(args pokesay.Args) {
	names := pokedex.ReadStructFromBytes[map[string][]int](GOBAllNames)
	timer.DebugTimer.Mark("read name struct")

	metadata, final := pokesay.ChooseByNameAndCategory(names, args.NameToken, GOBCowNames, MetadataRoot, args.Category)

	pokesay.Print(args, final.EntryIndex, GenerateNames(metadata, args, final), final.Categories, GOBCowData)
}

// runPrintRandom prints a random pokemon
// - This loads a specific GOB file from the embedded filesystem that contains the number of pokemon
// - generates a random number between 0 and the number of pokemon
// - reads the metadata file of at `<index>.metadata` as a PokemonMetadata struct
// - chooses a random entry from the metadata file
// - finally prints the pokemon
func runPrintRandom(args pokesay.Args) {
	choice := pokesay.RandomInt(pokedex.ReadIntFromBytes(GOBTotal))
	timer.DebugTimer.Mark("choose index")

	metadata := pokedex.ReadMetadataFromEmbedded(
		GOBCowNames,
		pokedex.MetadataFpath(MetadataRoot, choice),
	)

	final := metadata.Entries[pokesay.RandomInt(len(metadata.Entries))]
	timer.DebugTimer.Mark("choose entry")

	names := GenerateNames(metadata, args, final)
	timer.DebugTimer.Mark("generate names")

	pokesay.Print(args, final.EntryIndex, names, final.Categories, GOBCowData)
}

func main() {
	timer.DebugTimer.Mark("started main")

	args := parseFlags()
	timer.DebugTimer.Mark("parsed flags")

	// if the -h/--help flag is set, print usage and exit
	if args.Help {
		getopt.Usage()
		return
	}
	if args.Verbose {
		fmt.Println("Verbose output enabled")
		timer.DEBUG = true
	}

	if args.ListCategories {
		runListCategories()
	} else if args.ListNames {
		runListNames(args.ListNameToken)
	} else if args.NameToken != "" && args.Category != "" {
		runPrintByNameAndCategory(args)
	} else if args.NameToken != "" {
		runPrintByName(args)
	} else if args.IDToken != "" {
		runPrintByID(args)
	} else if args.Category != "" {
		runPrintByCategory(args)
	} else {
		runPrintRandom(args)
	}

	timer.DebugTimer.Mark("finish")

	timer.DebugTimer.Stop()
	timer.DebugTimer.PrintJson()
}
